sw/device/lib/dif/dif_usbdev.h

```diff
@@ -315,286 +344,283 @@ typedef enum dif_usbdev_rx_packet_get_info_result {
+ * host sends an IN token, clients must write the packet payload to a device
+ * buffer and mark it as ready for transmission from a particular endpoint. A
+ * packet queued for transmission from a particular endpoint is transmitted once
+ * the host sends an IN token for that endpoint.
+ *
+ * After a packet is queued for transmission, clients should check its status by
+ * calling `dif_usbdev_get_tx_status`. While the USB device handles transmission
+ * errors automatically by retrying transmission, transmission of a packet may
+ * be canceled if the endpoint receives a SETUP packet or the link is reset
+ * before the queued packet is transmitted. In these cases, clients should
+ * handle the SETUP packet or the link reset first and then optionally send the
+ * same packet again. Clients must also make sure that the given endpoint does
+ * not already have a packet pending for transmission before calling this
+ * function.
+ *
+ * See also: `dif_usbdev_buffer_request`, `dif_usbdev_buffer_write`,
+ * `dif_usbdev_get_tx_status`, `dif_usbdev_buffer_return`.
  *
  * @param usbdev A USB device.
  * @param endpoint An endpoint.
+ * @param buffer A buffer provided by `dif_usbdev_buffer_request`.
  * @return The result of the operation.
  */
 DIF_WARN_UNUSED_RESULT
-dif_usbdev_tx_packet_send_result_t dif_usbdev_tx_packet_send(
-    dif_usbdev_t *usbdev, uint8_t endpoint);
+dif_usbdev_result_t dif_usbdev_send(dif_usbdev_t *usbdev, uint8_t endpoint,
+                                    dif_usbdev_buffer_t *buffer);
 
 /**
  * Status of an outgoing packet.
  */
-typedef enum dif_usbdev_tx_packet_status {
+typedef enum dif_usbdev_tx_status {
   /**
    *  There is no packet for the given endpoint.
    */
-  kDifUsbdevTxPacketStatusNoPacket,
-  /**
-   * There is an active write buffer operation and the packet is not marked as
-   * ready for transmission yet.
-   */
-  kDifUsbdevTxPacketStatusStillWriting,
+  kDifUsbdevTxStatusNoPacket,
   /**
    * Packet is pending transmission.
    */
-  kDifUsbdevTxPacketStatusPending,
+  kDifUsbdevTxStatusPending,
   /**
    * Packet was sent successfully.
    */
-  kDifUsbdevTxPacketStatusSent,
+  kDifUsbdevTxStatusSent,
   /**
    * Transmission was canceled due to an incoming SETUP packet.
    */
-  kDifUsbdevTxPacketStatusCancelled,
-} dif_usbdev_tx_packet_status_t;
+  kDifUsbdevTxStatusCancelled,
+} dif_usbdev_tx_status_t;
 
 /**
  * Get the status of a packet that has been queued to be sent from an endpoint.
  *
+ * While the USB device handles transmission errors automatically by retrying
+ * transmission, transmission of a packet may be canceled if the endpoint
+ * receives a SETUP packet or the link is reset before the queued packet is
+ * transmitted. In these cases, clients should handle the SETUP packet or the
+ * link reset first and then optionally send the same packet again.
+ *
+ * This function returns the buffer that holds the packet payload to the free
```

Comment:
This doesn't need to be addressed in this PR, but if a user has attempted to send a packet, and got a SETUP, we should give the buffer back to them, rather than marking it as stale. This allows the user to decide if they want to discard that buffer or not, rather than the DIF just marking the buffer as stale. We can address this in a low-priority follow-up.

Reply:
I agree, added to #1601.